filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilder.scala

```diff
@@ -57,35 +59,71 @@ object BlobPathBuilder {
+  } yield nioPath) match {
+    case Success(value) => value
+    case Failure(exception: IOException) => exception.getCause() match {
+      // Azure NIO library wraps blobStorageExceptions in IOExceptions.
+      case cause: BlobStorageException => attempted match {
+        // If a restart of the filesystem was already attempted, throw the exception that the IO is wrapping
+        case true => throw cause
+        // This exception indicated that the filesystem was opened successfully, but something is wrong
+        // Try closing the filesystem, and opening with a fresh token
+        case false => {
+          closeFileSystem(buildURI(endpoint))
+          findNioPath(path, endpoint, container, blobTokenGenerator, true)
+        }
+      }
+      case _ => throw exception
     }
+    case Failure(exception) => throw exception
   }
 
-  def build(string: String): Try[BlobPath] = {
-    validateBlobPath(string, container, endpoint) match {
-      case ValidBlobPath(path) => for {
-            fileSystem <- retrieveFilesystem(new URI("azb://?endpoint=" + endpoint))
-            nioPath <- Try(fileSystem.getPath(path))
-            blobPath = BlobPath(nioPath, endpoint, container)
-          } yield blobPath
-      case UnparsableBlobPath(errorMessage: Throwable) => Failure(errorMessage)
+  def retrieveFilesystem(uri: URI, container: String, blobTokenGenerator: BlobTokenGenerator): Try[FileSystem] = {
+    Try(FileSystems.getFileSystem(uri)) recover {
+      // If no filesystem already exists, this will create a new connection, with the provided configs
+      case _: FileSystemNotFoundException => {
+        val fileSystemConfig = buildConfigMap(blobTokenGenerator.getAccessToken, container)
+        FileSystems.newFileSystem(uri, fileSystemConfig.asJava)
+      }
     }
   }
 
-  override def name: String = "Azure Blob Storage"
+  def closeFileSystem(uri: URI) = Try(FileSystems.getFileSystem(uri)).map(_.close)
+
+  def retrieveFilePath(fileSystem: FileSystem, path: String): NioPath = {
+    val blobPath = fileSystem.getPath(path)
+    fileSystem.provider().checkAccess(blobPath)
+    blobPath
+  }
 }
+case class BlobPath private[blob](pathString: String, endpoint: String, container: String, blobTokenGenerator: BlobTokenGenerator) extends Path {
+  //var token = blobTokenGenerator.getAccessToken
+  //var expiry = token.getSignature.split("&").filter(_.startsWith("se")).headOption.map(_.replaceFirst("se=",""))
+  override def nioPath: NioPath = BlobPath.findNioPath(path = pathString, endpoint, container, blobTokenGenerator)
 
-// Add args for container, storage account name
-case class BlobPath private[blob](nioPath: NioPath, endpoint: String, container: String) extends Path {
-  override protected def newPath(nioPath: NioPath): Path = BlobPath(nioPath, endpoint, container)
+  override protected def newPath(nioPath: NioPath): Path = BlobPath(nioPath.toString(), endpoint, container, blobTokenGenerator)
```

Comment:
Scala-ism: When calling no-arg Java accessors (like `toString`), we typically leave the parens off. In Scala, empty parens indicate that there *are* arguments, but we're using all default values instead of supplying them.